integrations/go/workload-executor.go

```diff
@@ -0,0 +1,309 @@
+}
+
+func verifyCursorResult(cur *mongo.Cursor, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	if cur == nil {
+		return false
+	}
+
+	defer func() {
+		err := cur.Close(context.Background())
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	for _, expected := range result.(bson.A) {
+		if !cur.Next(context.Background()) {
+			return false
+		}
+		if !bytes.Equal(expected.(bson.Raw), cur.Current) {
+			return false
+		}
+	}
+
+	if cur.Next(context.Background()) {
+		return false
+	}
+	return cur.Err() == nil
+}
+
+func verifyUpdateResult(res *mongo.UpdateResult, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	var expected struct {
+		MatchedCount  int64 `bson:"matchedCount"`
+		ModifiedCount int64 `bson:"modifiedCount"`
+		UpsertedCount int64 `bson:"upsertedCount"`
+	}
+	err := bson.Unmarshal(result.(bson.Raw), &expected)
+	if err != nil {
+		return false
+	}
+
+	if expected.MatchedCount != res.MatchedCount {
+		return false
+	}
+	if expected.ModifiedCount != res.ModifiedCount {
+		return false
+	}
+
+	actualUpsertedCount := int64(0)
+	if res.UpsertedID != nil {
+		actualUpsertedCount = 1
+	}
+	return expected.UpsertedCount == actualUpsertedCount
+}
+
+func executeCollectionOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	switch op.Name {
+	case "insertOne":
+		res, err := executeInsertOne(coll, op.Arguments)
+		return verifyInsertOneResult(res, op.Result), err
+	case "find":
+		cursor, err := executeFind(coll, op.Arguments)
+		return verifyCursorResult(cursor, op.Result), err
+	case "updateOne":
+		res, err := executeUpdateOne(coll, op.Arguments)
+		return verifyUpdateResult(res, op.Result), err
+	}
+	return false, errors.New("unrecognized collection operation: " + op.Name)
+}
+
+func runOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	// execute the command on the given object
+	if op.Object == "collection" {
+		return executeCollectionOperation(coll, op)
+	}
+	return false, errors.New("unrecognized object: " + op.Name)
+}
+
+func main() {
+	connstring := os.Args[1]
+	workloadSpec := os.Args[2]
+
+	var workload driverWorkload
+	err := bson.UnmarshalExtJSONWithRegistry(specTestRegistry, []byte(workloadSpec), false, &workload)
+	if err != nil {
+		panic(err)
+	}
+
+	client, err := mongo.Connect(context.Background(), options.Client().ApplyURI(connstring))
+	if err != nil {
+		panic(err)
+	}
+	defer func() { _ = client.Disconnect(context.Background()) }()
+
+	db := client.Database(workload.Database)
+	coll := db.Collection(workload.Collection)
+
+	results := struct {
+		NumErrors    int `json:"numErrors"`
+		NumFailures  int `json:"numFailures"`
+		NumSuccesses int `json:"numSuccesses"`
+	}{}
+
+	done := make(chan struct{})
+
+	// Waits for the termination signal from astrolabe and terminates the operation loop
+	go func() {
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```

Comment:
To check, this does not run on Windows, right? Otherwise, this may need to account for `CTRL_CLOSE_EVENT` as described in the Astrolabe docs https://mongodb-labs.github.io/drivers-atlas-testing/spec-workload-executor.html#behavioral-description

> The choice of termination signal used by astrolabe varies by platform. SIGINT 1 is used as the termination signal on Linux and OSX, while CTRL_BREAK_EVENT 2 is used on Windows.